pkg/proxy/nftables/proxier.go

```diff
@@ -0,0 +1,1631 @@
+	encoded := base32.StdEncoding.EncodeToString(hash[:])
+	name = encoded[:8] + "-" + name
+	if len(name) > chainNameBaseLengthMax {
+		name = name[:chainNameBaseLengthMax-3] + "..."
+	}
+	return name
+}
+
+func isServiceChainName(chainString string) bool {
+	// The chains returned from servicePortChainNameBase and
+	// servicePortEndpointChainNameBase will always have at least one "/" in them.
+	// (The chain prefix is at most 16 characters, the hash and surrounding hyphens
+	// add another 10, and the Namespace name is at most 63, meaning the "/" between
+	// Namespace and Name is at most 90 characters into the string, 12 characters
+	// before the truncation point. Since none of our "stock" chain names use slashes,
+	// we can distinguish them this way.
+
+	return strings.Contains(chainString, "/")
+}
+
+func isAffinitySetName(set string) bool {
+	return strings.HasPrefix(set, servicePortEndpointAffinityNamePrefix)
+}
+
+// This is where all of the nftables calls happen.
+// This assumes proxier.mu is NOT held
+func (proxier *Proxier) syncProxyRules() {
+	proxier.mu.Lock()
+	defer proxier.mu.Unlock()
+
+	// don't sync rules till we've received services and endpoints
+	if !proxier.isInitialized() {
+		klog.V(2).InfoS("Not syncing nftables until Services and Endpoints have been received from master")
+		return
+	}
+
+	//
+	// Below this point we will not return until we try to write the nftables rules.
+	//
+
+	// Keep track of how long syncs take.
+	start := time.Now()
+	defer func() {
+		metrics.SyncProxyRulesLatency.Observe(metrics.SinceInSeconds(start))
+		klog.V(2).InfoS("SyncProxyRules complete", "elapsed", time.Since(start))
+	}()
+
+	serviceUpdateResult := proxier.svcPortMap.Update(proxier.serviceChanges)
+	endpointUpdateResult := proxier.endpointsMap.Update(proxier.endpointsChanges)
+
+	klog.V(2).InfoS("Syncing nftables rules")
+
+	success := false
+	defer func() {
+		if !success {
+			klog.InfoS("Sync failed", "retryingTime", proxier.syncPeriod)
+			proxier.syncRunner.RetryAfter(proxier.syncPeriod)
+		}
+	}()
+
+	// If there are sufficiently-stale chains left over from previous transactions,
+	// try to delete them now.
+	if len(proxier.staleChains) > 0 {
+		oneSecondAgo := start.Add(-time.Second)
+		tx := proxier.nftables.NewTransaction()
+		deleted := 0
+		for chain, modtime := range proxier.staleChains {
+			if modtime.Before(oneSecondAgo) {
+				tx.Delete(&knftables.Chain{
+					Name: chain,
+				})
+				delete(proxier.staleChains, chain)
+				deleted++
+			}
+		}
+		if deleted > 0 {
+			klog.InfoS("Deleting stale nftables chains", "numChains", deleted)
+			err := proxier.nftables.Run(context.TODO(), tx)
+			if err != nil {
+				// We already deleted the entries from staleChains, but if
+				// the chains still exist, they'll just get added back
+				// (with a later timestamp) at the end of the sync.
+				klog.ErrorS(err, "Unable to delete stale chains; will retry later")
+				// FIXME: metric
+			}
+		}
+	}
```

Comment:
this is something unexpected to me, it seems worthy to have a metric to know how many times we hit this problem , of course it can be a follow up

The `1 time.Second` choice has any explanation or just a random value we choose?

Reply:
I think this is related to https://github.com/kubernetes/kubernetes/pull/121046#discussion_r1367954679